src/engagement_db_to_rapid_pro/engagement_db_to_rapid_pro.py

```diff
@@ -0,0 +1,79 @@
+    :param uuid_table: UUID table to use to de-identify contact urns.
+    :type uuid_table: id_infrastructure.firestore_uuid_table.FirestoreUuidTable
+    :param sync_config: Configuration for the sync.
+    :type sync_config: src.engagement_db_to_rapid_pro.configuration.EngagementDBToRapidProConfiguration
+    """
+    # Get all messages to sync.
+    # Only sync live messages. Anything else means we should signal Rapid Pro to request the data again if it can.
+    # TODO: This could get expensive as the number of datasets increases, in which case we can optimise by adding
+    #       an array-contains filter on the dataset field, so we only need to download messages in the datasets we
+    #       want to sync.
+    messages = engagement_db.get_messages(filter=lambda q: q.where("status", "in", [MessageStatuses.LIVE]))
+
+    # Organise messages by participant then by engagement db dataset.
+    participants = dict()  # of participant_uuid -> (dict of dataset -> list of Message)
+    for msg in messages:
+        if msg.participant_uuid not in participants:
+            participants[msg.participant_uuid] = dict()
+        participant = participants[msg.participant_uuid]
+
+        if msg.dataset not in participant:
+            participant[msg.dataset] = []
+        dataset = participant[msg.dataset]
+
+        dataset.append(msg)
+
+    # Make sure all the contact fields exist in the Rapid Pro workspace.
+    existing_contact_fields = [f.key for f in rapid_pro.get_fields()]
+    contact_fields_to_sync = [dataset_config.rapid_pro_contact_field for dataset_config in sync_config.normal_datasets]
+    for contact_field in contact_fields_to_sync:
+        if contact_field not in existing_contact_fields:
+            rapid_pro.create_field(contact_field)
+
+    # Sync each participant to Rapid Pro.
+    for i, (participant_uuid, datasets) in enumerate(participants.items()):
+        log.info(f"Syncing participant {i + 1}/{len(participants)}: {participant_uuid}...")
+        # Re-identify the participant
+        urn = uuid_table.uuid_to_data(participant_uuid)
+
+        # Build a dictionary of contact_field -> value to write for each normal dataset.
+        contact_fields = dict()
+        for dataset_config in sync_config.normal_datasets:
+            # Find all the messages from this participant that are relevant to this dataset.
+            messages = []
+            for dataset in dataset_config.engagement_db_datasets:
+                messages.extend(datasets.get(dataset, []))
+
+            # Only overwrite this contact field if there is data to write or it's ok to clear a field.
+            if len(messages) > 0:
+                if sync_config.write_mode == WriteModes.SHOW_PRESENCE:
+                    contact_fields[dataset_config.rapid_pro_contact_field] = _PRESENCE_VALUE
+                else:
+                    assert sync_config.write_mode == WriteModes.CONCATENATE_TEXTS
```

Comment:
Which instance would we want to concatenate texts in rapidpro.

Reply:
I think it's a useful debug tool; it's much easier to understand/check what's going on in rapid pro if you can see what the messages were

Reply:
Maybe am missing something here - but I think we should not be updating texts in fields that also store direct responses from users. e.g storing age response from Imaqal to a csap age responce field.

Reply:
resolved offline to include origins in contact field write-backs in a future pr